mmaction/models/recognizers/recognizer3d.py

```diff
@@ -38,22 +41,46 @@ def _do_test(self, imgs):
                 'max_testing_views is only compatible '
                 'with batch_size == 1')
             view_ptr = 0
-            cls_scores = []
+            feats = []
             while view_ptr < total_views:
                 batch_imgs = imgs[view_ptr:view_ptr + self.max_testing_views]
                 x = self.extract_feat(batch_imgs)
                 if self.with_neck:
                     x, _ = self.neck(x)
-                cls_score = self.cls_head(x)
-                cls_scores.append(cls_score)
+                feats.append(x)
                 view_ptr += self.max_testing_views
-            cls_score = torch.cat(cls_scores)
+            # should consider the case that feat is a tuple
+            if isinstance(feats[0], tuple):
+                len_tuple = len(feats[0])
+                feat = [
+                    torch.cat([x[i] for x in feats]) for i in range(len_tuple)
+                ]
+                feat = tuple(feat)
+            else:
+                feat = torch.cat(feats)
         else:
-            x = self.extract_feat(imgs)
+            feat = self.extract_feat(imgs)
             if self.with_neck:
-                x, _ = self.neck(x)
-            cls_score = self.cls_head(x)
-
+                feat, _ = self.neck(feat)
+
+        if self.feature_extraction:
+            # perform spatio-temporal pooling
+            avg_pool = nn.AdaptiveAvgPool3d(1)
```

Comment:
F.adaptive_avg_pool3d

Reply:
I guess there is little performance difference 😄